### PR TITLE
GH-2143 ensure project is binary compatible with 3.1 releases

### PR DESCRIPTION
**Pull request by jeenbroekstra**
GitHub issue resolved: #2143  

Briefly describe the changes proposed in this PR:

- modified several RDF* related changes to ensure reasonable defaults are in place
- modified several recent public and protected method changes to ensure binary compatibility
- configured [japicmp](https://siom79.github.io/japicmp/MavenPlugin.html) to check (and fail!) build on both source and binary compatibility with previous minor release. It ignore items annotated with InternalUseOnly or Experimental.  

---- 
PR Author Checklist: 

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [x] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/QueryJoinOptimizer.java [hunk 1]
```diff
@@ -393,6 +393,12 @@ protected TupleExpr selectNextTupleExpr(List<TupleExpr> expressions, Map<TupleEx
 			return result;
 		}
 
+		@Deprecated
+		protected double getTupleExprCardinality(TupleExpr tupleExpr, Map<TupleExpr, Double> cardinalityMap,
```
Comment (jeenbroekstra) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/QueryJoinOptimizer.java:397]: This is one of the many reasons why I want stop using `protected` as a modifier, as much as possible.

**Comment by jeenbroekstra** [comment]
> I do not have a good understanding of the Maven plugin, can you maybe shortly comment how it works and how it influences the build?

Sure thing. It ties in at the `verify` phase of the maven build cycle, and what it does is a binary comparison of each jar artifact, comparing the current version against a specified old version (in this case, 3.1.4) for compatibility. I've configured it to detect both binary and source incompatibilities and to fail the build on both (so we get a definite warning) - this will effectively mean that pull requests will be blocked until compatibility issues are resolved. 

It currently checks compatibility for everything that is public or protected, with the exception of anything that is annotated as `@Experimental` or `@InternalUseOnly`. I had to tweak a bit to not flag certain corner cases in our code base, and I don't claim that the current setup is 100% correct and guarantees backward compatibilty, but it definitely flags most common issues and that can only be helpful.

**Comment by hmottestad** [comment]
I think this looks good :) It will probably cause some hassle in the beginning as we discover if it manages to exclude everything we need. Could you create a separate build stage in the Github Actions just for this part. Something like this:

1. verify formatting
1. verify that code compiles
1. verify that code is source and binary compatible
1. verify that tests pass

Also: Do you need to bump the version number up for every release?

**Comment by jeenbroekstra** [comment]
Good point, I'll have a quick look at configuring github actions. 

> Also: Do you need to bump the version number up for every release?

Not necessarily - if we commit to all minor releases being binary compatible, we can keep comparing with the currently set version for every future minor release. Of course, we do need to bump it as soon as we start working on the next major.

**Pull request merged**